matrix-synapse: 0.34.1.1 -> 0.99.0

nyanloutre opened the pull request:
###### Motivation for this change

There is a new optional dependency to support automatic ACME certificate, maybe we should add this

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

nyanloutre committed:
pythonPackages.pymacaroons: init at 0.13.0

nyanloutre committed:
matrix-synapse: 0.34.1.1 -> 0.99.0

Also cleanup of old dependencies and irrelevant patch

nyanloutre committed:
pythonPackages.pymacaroons-pynacl: remove unmaintained fork

Mic92 commented:
Any details what is wrong with the tests at the moment?

nyanloutre commented:
@Mic92 it seems `/bin/homeserver ` no longer exists and `python -m synapse.app.homeserver` is used instead

Mic92 commented:
@nyanloutre ok. so our nixos service also needs to be adapted.
I mark this as WIP.

nyanloutre committed:
nixos/tests/matrix-synapse: generate ca and certificates

nyanloutre commented:
I had to rewrote the test a bit because synapse will not generate self signed certificates anymore.
I did something similar to the etcd-cluster test where a CA and a rsa key are generated.

nyanloutre committed:
nixos/matrix-synapse: use python to launch synapse

launch synapse with the python executable because the startup script is
no longer available

nyanloutre commented:
@Mic92 my PR should be ready to merge, what do you think ?

florianjacob commented:
While we are at updating the NixOS service, as synapse will require a valid certificate soon, might be a good idea to add the new [certificate reloading via SIGHUP](https://github.com/matrix-org/synapse/blob/master/docs/MSC1711_certificates_FAQ.md#how-do-i-tell-synapse-to-reload-my-keyscertificates-after-i-replace-them) as `ExecReload` for the service for easy use with letsencrypt.

dotlambda commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -651,12 +651,16 @@ in {
       wantedBy = [ "multi-user.target" ];
       preStart = ''
-        ${cfg.package}/bin/homeserver \
+        ${python}/bin/python -m synapse.app.homeserver \
```
```suggestion
        ${python.interpreter} -m synapse.app.homeserver \
```

dotlambda commented on nixos/tests/matrix-synapse.nix:
```diff
@@ -8,23 +36,31 @@ import ./make-test.nix ({ pkgs, ... } : {
+      services.matrix-synapse = {
+        enable = true;
+        database_type = "psycopg2";
+        tls_certificate_path = "${cert}";
```
```suggestion
        tls_certificate_path = cert";
```

nyanloutre replied:
I got an error with this as `cert` is a path and `tls_certificate_path` wants a string

dotlambda replied:
Makes sense :)

nyanloutre committed:
nixos/matrix-synapse: reload service with SIGHUP

This is used to load new certificates without restarting the service

pstn commented:
I think this needs to be backported to stable. They announcde that they are going to [break downwards compatiblity](https://matrix.org/blog/2019/02/05/synapse-0-99-0/) in one month. That's too soon to wait for nixos-19.03

Ekleog reviewed (approved):
LGTM! I'll leave it here a bit for @Ralith to check.

However, I don't think we should backport this: it breaks backwards-compatibility in a way that will require manual intervention. Meaning that if we backport and the user does nothing the user no longer has any server, while if we don't backport and the user does nothing then the user has a server that just can't federate with too-up-to-date servers.

And the upgrade is not actually required to be done by 1 month. The only thing that is really required is to have a non-self-signed certificate. Which can perfectly well be done from 0.34.1.1, and which would have to be done manually anyway, as it's the steps for migrating to 0.99.0.

So overall I think we shouldn't backport such a heavily-backwards-compatibility-breaking update.

As an aside, I think the 1-month delay is stupid, but that is not my choice to make and I already voiced it upstream.

Ekleog commented:
Oh, and…

@GrahamcOfBorg build matrix-synapse nixosTests.matrix-synapse

pstn commented:
What about backporting it to a new package? Then users would have a choice of having the pain now or just waiting for 19.03.

Ralith commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -687,7 +691,7 @@ in {
         PermissionsStartOnly = true;
         ExecStart = ''
-          ${cfg.package}/bin/homeserver \
+          ${python.interpreter} -m synapse.app.homeserver \
```
This should probably be squashed into the commit that removed the script.

nyanloutre replied:
what commit do you mean ?

Ralith reviewed (commented):
Great cleanup, thanks!

dotlambda commented:
> What about backporting it to a new package? Then users would have a choice of having the pain now or just waiting for 19.03.

What's the advantage over simply using the package from the unstable channel on a stable system?

7c6f434c commented:
>What's the advantage over simply using the package from the unstable channel on a stable system?

Not having the second glibc and python in the store, I guess.

Re: migration pain — as far as I know, Synapse 0.99 can be told to obtain ACME certificates for itself https://github.com/matrix-org/synapse/blob/3bd9daf4b86ed811c9ced95a4adb9aa58b681399/docs/ACME.md

nyanloutre commented:
@7c6f434c for automatic ACME to work it needs an optional library that is not yet packaged in nixpkgs

florianjacob commented:
Note that 0.99 and 1.0 can still talk to 0.34.1 if 0.34.1 is equipped with a valid certificate, so while manual intervention is always required, the backport is not actually necessary to stay compatible, as you can use the usual nixos machinery for getting certificates.

This is roughly what I use with 0.34.1 to stay compatible with 0.99 and 1.0, without actually needing to update to 0.99 yet. While I'll personally updated to 0.99 by using the unstable package & module, this code is *the same* what you will need to do for 0.99 anyway, if you don't plan to use the integrated ACME support or want to use a tls terminating reverse proxy for port 8448.

```
services.matrix-synapse = {
    …
    tls_private_key_path = "${config.security.acme.directory}/${config.networking.hostName}.${config.networking.domain}/key.pem";                                                    
    tls_certificate_path = "${config.security.acme.directory}/${config.networking.hostName}.${config.networking.domain}/fullchain.pem";
};

# group used for permission to get and use the fqdn certificate (in contrast to subdomains of this host)
users.groups.domaincertificate = {};
users.users."matrix-synapse".extraGroups = [ "domaincertificate" ];
users.users."nginx".extraGroups = [ "domaincertificate" ];

services.nginx = {
  enable = true;
  virtualHosts = {
    "${config.networking.hostName}.${config.networking.domain}" = {
      default = true;
      enableACME = true;
  };
};
security.acme.certs."${config.networking.hostName}.${config.networking.domain}" = {
    group = "domaincertificate";
    allowKeysForGroup = true;
    # TODO: use reload: https://github.com/NixOS/nixpkgs/pull/55320#issuecomment-461056137
    postRun = "systemctl restart matrix-synapse.service";
};
```

Ekleog commented:
@florianjacob Could you consider submitting this code as a PR to the NixOS manual? It sounds like it'd be pretty useful for people :)

@nyanloutre Upon second review I think this is missing release notes (nixos/doc/manual/release-notes/rl-1903.xml). Could you consider adding a line there pointing to the need for a non-self-signed certificate? Then @florianjacob's PR, if he feels like it, would be able to add a link to this line pointing to the NixOS manual section for how to actually do it.

nyanloutre committed:
nixos/release-notes: mention breaking changes with matrix-synapse update

nyanloutre commented:
@Ekleog done

Ekleog commented:
@nyanloutre Thank you! Merged :)

@pstn Backporting to a separate package sounds against the spirit of `release-18.09` to me, especially as there is no actual need to upgrade to keep federation working. However, feel free to add the updated package to your [NUR](https://github.com/nix-community/NUR) for others to be able to use, if you really want to stay on 18.09 but upgrade without taking the package from unstable :)

@florianjacob If you decide to send your configuration excerpt as a PR to the manual and add a link to it from the release notes, feel free to ping me. :)

Ralith commented:
@florianjacob You might be interested in the `systemctl reload-or-restart` action, which just does the right thing here.

pacien commented:
I've attempted running this on NixOS 18.09 by importing the module and package form the unstable channel.

The service refuses to start with the following error due to a Python version mismatch (probably in a similar way as in https://github.com/matrix-org/synapse/issues/4431):
```
Feb 09 06:37:50 matrix anw2ll0dqfkyql2gmfhsi4ccwc212nrp-unit-script-matrix-synapse-pre-start[451]: /nix/store/99p0nyqgayic2ipqx2x4lr4njl8f5nrz-python3-3.6.8/bin/python3.6m: Error while finding module specification for 'synapse.app.homeserver' (ModuleNotFoundError: No module named 'synapse')
```

Setting the version properly in the service configuration [here](https://github.com/nixos/nixpkgs/blob/e088eb34d93b0adcdfd46adf7eb7c35bcde346fc/nixos/modules/services/misc/matrix-synapse.nix#L656) allows the `synapse` python module to be found, but not its runtime dependencies:
```
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]: Traceback (most recent call last):
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:   File "/nix/store/5jy6nkhccr8j2bycl6qi404zwnq0swak-python3-3.7.2-env/lib/python3.7/runpy.py", line 183, in _run_module_as_main
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:     mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:   File "/nix/store/5jy6nkhccr8j2bycl6qi404zwnq0swak-python3-3.7.2-env/lib/python3.7/runpy.py", line 109, in _get_module_details
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:     __import__(pkg_name)
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:   File "/nix/store/5jy6nkhccr8j2bycl6qi404zwnq0swak-python3-3.7.2-env/lib/python3.7/site-packages/synapse/app/__init__.py", line 18, in <module>
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:     from synapse import python_dependencies  # noqa: E402
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:   File "/nix/store/5jy6nkhccr8j2bycl6qi404zwnq0swak-python3-3.7.2-env/lib/python3.7/site-packages/synapse/python_dependencies.py", line 19, in <module>
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]:     from pkg_resources import DistributionNotFound, VersionConflict, get_distribution
Feb 09 06:43:33 matrix jk958mp32snk3rdx9xf8yjslk8prbxim-unit-script-matrix-synapse-pre-start[665]: ModuleNotFoundError: No module named 'pkg_resources'
```

dotlambda commented:
@pacien Try this: https://github.com/dotlambda/nixpkgs/commits/matrix-synapse-correct-python. It might be a little overcomplicated but I think it's the correct way to do this.

pacien commented:
@dotlambda This works great, thanks!
Though wouldn't it be simpler to re-create `bin/homeserver` as a wrapper in the synapse package to bind the right interpreter and environment to the application instead of exposing the internal interpreter?

dotlambda commented:
> Though wouldn't it be simpler to re-create `bin/homeserver` as a wrapper in the synapse package to bind the right interpreter and environment to the application instead of exposing the internal interpreter?

It would and I have no clue what hinders upstream from doing this.

florianjacob commented:
@Ekleog working on a manual section, I actually now switched to a different, easier and more robust setup.

@dotlambda your changes seem very sensible to me, the nixos module shouldn't use whatever is in `pkgs.python` but the actual python from `matrix-synapse.package`, I started using them as well to test synapse from unstable on 18.09. Do you intend to upstream them?

dotlambda commented:
> @dotlambda your changes seem very sensible to me, the nixos module shouldn't use whatever is in `pkgs.python` but the actual python from `matrix-synapse.package`, I started using them as well to test synapse from unstable on 18.09. Do you intend to upstream them?

I leave that up to the module maintainers.

pacien commented on pkgs/servers/matrix-synapse/matrix-synapse.patch:
```diff
@@ -1,20 +0,0 @@
-     zip_safe=False,
-     long_description=long_description,
--    scripts=["synctl"] + glob.glob("scripts/*"),
-+    scripts=["synctl", "homeserver"] + glob.glob("scripts/*"),
```
@nyanloutre What was the reason for removing this?

nyanloutre replied:
This is because there isn't any `homeserver` script anymore

pacien replied:
The patch was actually creating this wrapper script.
I've opened a PR to restore it.

nyanloutre replied:
But still, it's better if we can avoid a patch

dotlambda replied:
Is there a reason why this patch can't be upstreamed? I'd much prefer the patch to my (more complicated) solution.

nyanloutre replied:
https://github.com/matrix-org/synapse/issues/1075

nyanloutre commented:
Le 23 février 2019 21:08:42 GMT+01:00, Notkea <notifications@github.com> a écrit :
>pacien commented on this pull request.>
>>
>>
>>
>> -index 0000000..2f1d413>
>---- /dev/null>
>-+++ b/homeserver>
>-@@ -0,0 +1,1 @@>
>-+synapse/app/homeserver.py>
>-\ No newline at end of file>
>-diff --git a/setup.py b/setup.py>
>-index b00c2af..c7f6e0a 100755>
>---- a/setup.py>
>-+++ b/setup.py>
>-@@ -92,6 +92,6 @@ setup(>
>-     include_package_data=True,>
>-     zip_safe=False,>
>-     long_description=long_description,>
>--    scripts=["synctl"] + glob.glob("scripts/*"),>
>-+    scripts=["synctl", "homeserver"] + glob.glob("scripts/*"),>
>>
>The patch was actually creating this wrapper script.>
>I've opened a PR to restore it.>
>>
>-- >
>You are receiving this because you were mentioned.>
>Reply to this email directly or view it on GitHub:>
>https://github.com/NixOS/nixpkgs/pull/55320#discussion_r259591494

But still, it's better if we can avoid a patch
-- 
Paul TREHIOU

florianjacob commented:
@Ekleog Guide is finished: https://github.com/NixOS/nixpkgs/pull/57699
